src/main/java/com/cburch/logisim/gui/start/GuiInterface.java

```diff
@@ -0,0 +1,285 @@
+/*
+ * Logisim-evolution - digital logic design tool and simulator
+ * Copyright by the Logisim-evolution developers
+ *
+ * https://github.com/logisim-evolution/
+ *
+ * This is free software released under GNU GPLv3 license
+ */
+
+package com.cburch.logisim.gui.start;
+
+import java.awt.AWTEvent;
+import java.awt.Component;
+import java.awt.Point;
+import java.awt.Toolkit;
+import java.awt.event.AWTEventListener;
+import java.awt.event.ContainerEvent;
+import java.io.File;
+import java.util.ArrayList;
+
+import javax.help.JHelp;
+import javax.swing.JButton;
+import javax.swing.JCheckBox;
+import javax.swing.JCheckBoxMenuItem;
+import javax.swing.JComboBox;
+import javax.swing.JComponent;
+import javax.swing.JFileChooser;
+import javax.swing.JFrame;
+import javax.swing.JLabel;
+import javax.swing.JMenuItem;
+import javax.swing.JOptionPane;
+import javax.swing.JPanel;
+import javax.swing.JProgressBar;
+import javax.swing.JRadioButton;
+import javax.swing.JRadioButtonMenuItem;
+import javax.swing.JScrollPane;
+import javax.swing.JSpinner;
+import javax.swing.JTabbedPane;
+import javax.swing.JTextArea;
+import javax.swing.JTextField;
+import javax.swing.JToolTip;
+import javax.swing.UIDefaults;
+import javax.swing.UIManager;
+
+import com.cburch.logisim.file.LoadFailedException;
+import com.cburch.logisim.file.Loader;
+import com.cburch.logisim.gui.generic.CanvasPane;
+import com.cburch.logisim.gui.generic.OptionPane;
+import com.cburch.logisim.gui.icons.ErrorIcon;
+import com.cburch.logisim.gui.icons.InfoIcon;
+import com.cburch.logisim.gui.icons.QuestionIcon;
+import com.cburch.logisim.gui.icons.WarningIcon;
+import com.cburch.logisim.gui.main.Print;
+import com.cburch.logisim.gui.menu.LogisimMenuBar;
+import com.cburch.logisim.gui.menu.WindowManagers;
+import com.cburch.logisim.prefs.AppPreferences;
+import com.cburch.logisim.proj.ProjectActions;
+import com.cburch.logisim.std.base.BaseLibrary;
+import com.cburch.logisim.std.gates.GatesLibrary;
+import com.cburch.logisim.util.MacCompatibility;
+
+public class GuiInterface implements AWTEventListener {
+
+  private static ArrayList<File> filesToOpen;
+  private static ArrayList<File> filesToPrint = new ArrayList<File>();
+  private static boolean initialized = false;
+
+  private String gateShape = null;
+  private Point windowSize = null;
+  private Point windowLocation = null;
+  private File templateFile = null;
+  private int templateType = AppPreferences.TEMPLATE_UNKNOWN;
+  private boolean showSplash = true;
+  private boolean clearPreferences = false;
+
+  public GuiInterface(Startup startup) {
+    assert startup.task == Startup.Task.GUI;
```

Comment:
Please no `assert` outside tests (multiple places).

Reply:
Why not? I have no problem if asserts get used in production code to document preconditions and postconditions of functions, which are supposed to always hold. I completely agree that assert is not the proper way to handle runtime errors, which can occur during normal operation. These need to be properly handled to allow either continuation of the program or exiting cleanly the application with meaningful error messages and an appropriate return code.

Reply:
@maehne: The reason is that assert is an old technique that predates object oriented programming.

The better solution for checking the validity of parameters that should never be invalid, is to do a test and throw an IllegalArgumentException. That way you can give a message about why the parameters are invalid and the report includes information about how you were called. It also plays nice with junit tests. That is, it is possible to write junit tests that test whether the validity check works.

Note that the IllegalArgumentException is not a message to the user, but rather to the programmer who incorrectly called your method. The Java libraries use it extensively. They also use the IllegalStateException when the current object state does not allow an operation to continue, such as removing an object from an empty queue.

In some of the cases here, @dtowell may be using assert as debugging code and they should probably be removed.

Reply:
Thanks @davidhutchens for pointing out the canonic way for handling preconditions in Java. It shows that it is not my primary (and most loved) programming language…

Reply:
@maehne couple of reasons for "why not" really, with first, and foremost being the fact the `assert` is **disabled** by default. So the `assert` will in fact do nothing at runtime. Not seasoned devs tend to fail into that trap pretty often. 2nd, it's extremely bad U/X to just kill the whole app and throw verbose stacktrace at Joe user who can have no clue WTF it even is and why this crappy app crashed on him. If the assert is triggered by user action, then s/he should be informed on why or what happened instead by dialog or any other form. Therefore `assert` should be part of your test code because you actually want them crash as verbose as possible. Once you done with tests, you should have all the cases covered in the code anyway, so again. it'd serve no purpose (yes, I know LSe and tests are not yet married really, but that's another, independent story).

Reply:
I will convert the asserts to throw IllegalArgumentException. (There are existing tests for these cases already, the assertions were not begin used to debug but to enlighten future programmers, but we can use exceptions as @davidhutchens points out.)